src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/CodeView/CodeViewFileTableBuilder.cs

```diff
@@ -0,0 +1,67 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Buffers;
+using System.Buffers.Binary;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+
+using static ILCompiler.ObjectWriter.CodeViewNative;
+
+namespace ILCompiler.ObjectWriter
+{
+    internal sealed class CodeViewFileTableBuilder
+    {
+        private MemoryStream _stringTableWriter = new();
+        private MemoryStream _fileTableWriter = new();
+        private Dictionary<string, uint> _fileNameToIndex = new();
+
+        public CodeViewFileTableBuilder()
+        {
+            // Insert empty entry at the beginning of string table
+            _stringTableWriter.Write(stackalloc byte[2]);
+        }
+
+        public uint GetFileIndex(string fileName)
+        {
+            if (fileName == "")
+            {
+                fileName = "<stdin>";
```

Comment:
It mimics what [LLVM uses](https://github.com/dotnet/llvm-project/blob/e80993c0649d5e949eb4ac23fe22290b4ecb7e66/llvm/lib/MC/MCCodeView.cpp#L59) (on multiple places). Some other LLVM tools do string based checks on it.

Empty file names have to be special-cased to prevent the UTF-8 string `\0` terminator being treated as the end of the file name list.